File: daemon/src/usage.js
// Real usage, straight from Claude Code.
//
// `claude -p "/usage"` runs the slash command locally and prints the same
// figures the interactive /usage view shows — percentages against the actual
// plan limits, reset times, and the "what's contributing" breakdown. It costs
// no model tokens (no inference happens) and, with
// CLAUDE_CODE_SKIP_PROMPT_HISTORY=1, writes no transcript.
//
// Everything here is parsed from that text. We never invent a denominator.

import crypto from 'node:crypto';
import os from 'node:os';
import { execFile } from 'node:child_process';
import { USAGE_REFRESH_MS } from './config.js';
import { markOwnSession } from './own-sessions.js';
import { readState, writeState } from './persist.js';
import { log } from './log.js';

// Where the last good reading is kept between runs.
const STATE_NAME = 'usage';

// A run costs no inference but still boots Claude Code and reads the plan, which
// takes the better part of 20 seconds on a cold start. The old 25s ceiling cut
// slower runs off, and the timeout then reported whatever stderr happened to
// hold — which is how a harmless stdin warning ended up on the usage screen as
// the failure. Well under the 60s refresh, so runs still never overlap.
const RUN_TIMEOUT_MS = 45_000;

// "Current session: 11% used · resets Jul 30 at 5:19am (America/New_York)"
// — the CLI's older format. Kept for backward compatibility; CLI 2.1.x+ uses
// the progress-bar block matched by PROGRESS_BAR_RE below instead.
const LIMIT_RE = /^\s*Current\s+(session|week[^:]*):\s*(\d+)%\s*used(?:\s*·\s*resets\s+([^(\n]+?))?\s*(?:\(([^)]+)\))?\s*$/i;
// "Last 24h · 579 requests · 8 sessions" — older format's window header.
const WINDOW_RE = /^\s*Last\s+(\S+)\s*·\s*([\d,]+)\s+requests\s*·\s*([\d,]+)\s+sessions\s*$/i;
// "Top skills: /webapp-testing 4%, /frontend-design 1%" — older format.
const TOP_RE = /^Top\s+(skills|subagents|MCP servers):\s*(.+)$/i;

// Current CLI: a rendered progress bar (block-drawing chars) ending "NN% used",
// preceded by a heading line ("Claude Code and Cowork credit") and followed by
// a detail line ("One-time credit · Expires September 29"). The heading isn't
// reliable across plan types, so it's read from whatever precedes the bar.
const PROGRESS_BAR_RE = /(\d+)%\s*used\s*$/;
// "Skills                  % of usage" / "Subagents ..." / "MCP servers ..."
const TABLE_HEADER_RE = /^\s*(Skills|Subagents|MCP servers)\s{2,}%\s*of\s*usage\s*$/i;
// "<name>          <NN>%" — at least two spaces before the trailing percentage
// (the columns are padded to align), so a stray prose line can't be mistaken
// for a row — those never end in a bare "NN%" here.
const TABLE_ROW_RE = /^\s*(\S.*?)\s{2,}(\d+)%\s*$/;
// "Showing last-known usage as of 2m ago (rate limited — try again in a moment)"
const STALE_NOTICE_RE = /^\s*Showing\s+last-known\s+usage\s+as\s+of\s+(.+?)\s+ago/i;

// Trailing percentage per item; anything that doesn't match that shape is kept
// with an empty value rather than dropped, so an unparsed entry still shows.
function parseTopList(rest) {
  return String(rest).split(',').map((chunk) => {
    const item = chunk.trim();
    const m = /^(.*\S)\s+(\d+%)$/.exec(item);
    return m ? { name: m[1], pct: m[2] } : { name: item, pct: '' };
  }).filter((x) => x.name);
}

function labelFor(kind) {
  const k = kind.toLowerCase();
  if (k === 'session') return 'SESSION';
  const model = /week\s*\(([^)]+)\)/i.exec(kind);
  if (model) {
    const name = model[1].trim();
    return /all models/i.test(name) ? 'WEEK · ALL MODELS' : `WEEK · ${name.toUpperCase()}`;
  }
  return kind.toUpperCase();
}

export function parseUsage(text, now = Date.now()) {
  const lines = String(text).split('\n').map((l) => l.replace(/\s+$/, ''));
  const rawLines = String(text).split('\n');
  const limits = [];
  const windows = [];
  let current = null;
  let subscription = null;
  let stale = false;
  // Which table a run of rows belongs to — only set right after a table
  // header, cleared on any line that doesn't look like a row (esp. blanks).
  let inTable = null;

  function ensureWindow() {
    if (!current) {
      current = { window: 'Last 24h', requests: 0, sessions: 0, notes: [], skills: [], subagents: [], mcp: [] };
      windows.push(current);
    }
  }

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const raw = rawLines[i];
    if (!line.trim()) { inTable = null; continue; }

    if (/subscription to power your Claude Code usage/i.test(line)) {
      subscription = line.trim();
      continue;
    }

    if (STALE_NOTICE_RE.test(line)) {
      stale = true;
      continue;
    }

    const limit = LIMIT_RE.exec(line);
    if (limit) {
      limits.push({
        key: limit[1].toLowerCase().replace(/[^a-z]+/g, '-'),
        label: labelFor(limit[1]),
        used: Number(limit[2]) / 100,
        detail: limit[3] ? `resets ${limit[3].trim()}` : '',
      });
      continue;
    }

    const win = WINDOW_RE.exec(line);
    if (win) {
      current = {
        window: `Last ${win[1]}`,
        requests: Number(win[2].replace(/,/g, '')),
        sessions: Number(win[3].replace(/,/g, '')),
        notes: [], skills: [], subagents: [], mcp: [],
      };
      windows.push(current);
      continue;
    }

    // Current-format progress-bar block: heading line (previous), the bar
    // itself (this line), optional detail line (next).
    const bar = PROGRESS_BAR_RE.exec(line);
    if (bar) {
      const heading = (i > 0 ? lines[i - 1].trim() : '') || 'Usage';
      const nextLine = i + 1 < lines.length ? lines[i + 1].trim() : '';
      const detail = nextLine && !TABLE_HEADER_RE.test(nextLine) && !PROGRESS_BAR_RE.test(nextLine) && !LIMIT_RE.test(nextLine)
        ? nextLine : '';
      limits.push({
        key: heading.toLowerCase().replace(/[^a-z]+/g, '-').replace(/^-+|-+$/g, ''),
        label: heading.toUpperCase(),
        used: Number(bar[1]) / 100,
        detail,
      });
      continue;
    }

    const tableHeader = TABLE_HEADER_RE.exec(line);
    if (tableHeader) {
      const key = { skills: 'skills', subagents: 'subagents', 'mcp servers': 'mcp' }[tableHeader[1].toLowerCase()];
      if (key) {
        ensureWindow();
        inTable = key;
      }
      continue;
    }

    if (inTable) {
      const row = TABLE_ROW_RE.exec(line);
      if (row) {
        current[inTable].push({ name: row[1].trim(), pct: `${row[2]}%` });
        continue;
      }
      inTable = null;
    }

    // Old-format indented bullets under a window ("  Top skills: a 4%, b 1%").
    if (current && /^\s{2,}\S/.test(raw) && line.trim()) {
      const note = line.trim();
      current.notes.push(note);
      const top = TOP_RE.exec(note);
      if (top) {
        const key = { skills: 'skills', subagents: 'subagents', 'mcp servers': 'mcp' }[top[1].toLowerCase()];
        if (key) current[key] = parseTopList(top[2]);
      }
    }
  }

  if (!limits.length) return null;

  const result = {
    updatedTs: now,
    updatedLabel: 'updated ' + new Date(now).toTimeString().slice(0, 5) + ' · from claude /usage',
    subscription,
    limits,
    windows,
  };
  if (stale) result.stale = true;
  return result;
}

// --- keeping a reading honest across polls ------------------------------------
//
// `claude -p /usage` does not always ask the server. It prints whatever is in
// `cachedUsageUtilization` in ~/.claude.json, and that file is read-modify-
// written wholesale by every claude process on the machine. A long-lived
// session that loaded it an hour ago writes its own stale copy back over the
// fresh one, so consecutive polls a minute apart can disagree — the screen
// flips between the real figure and a stale lower one until something settles.
// Observed live: 1% used, then 0% used, inside the same five-hour window.
//
// The fix is the one fact the reading itself gives us: usage inside a window
// only ever goes up. A lower number for the same window is a stale read, not a
// refund, so the higher one stands. When the window really does roll over the
// reset clause changes with it, and the drop is taken at face value.

function sameWindow(prev, next) {
  // A reading with no reset clause carries no window identity of its own — the
  // zeroed-out shape a clobbered cache prints — so it cannot claim to be a new
  // window. Only a different, stated reset time counts as a rollover.
  if (!prev.detail || !next.detail) return true;
  return prev.detail === next.detail;
}

export function reconcileUsage(prev, next) {
  if (!next || !Array.isArray(next.limits)) return next;
  const before = new Map(((prev && prev.limits) || []).map((l) => [l.key, l]));
  const limits = next.limits.map((l) => {
    const p = before.get(l.key);
    if (!p || !(p.used > l.used) || !sameWindow(p, l)) return l;
    // Held: the previous reading wins, and keeps its reset clause if this one
    // arrived without any.
    return { ...l, used: p.used, detail: l.detail || p.detail };
  });
  return { ...next, limits };
}

// The last good reading, so a restart shows real figures instead of spending a
// minute on "READING USAGE…". Flagged stale until the first live poll lands.
function loadPersisted() {
  const saved = readState(STATE_NAME);
  if (!saved || !Array.isArray(saved.limits) || !saved.limits.length) return null;
  const at = saved.updatedTs ? new Date(saved.updatedTs).toTimeString().slice(0, 5) : '';
  return {
    ...saved,
    stale: true,
    error: undefined,
    updatedLabel: `last reading${at ? ' ' + at : ''} · from claude /usage`,
  };
}

// What actually went wrong, in the words the device has room for. A killed run
// timed out — say that, rather than quoting whichever line stderr happened to
// end on, which is how warnings get mistaken for causes.
export function describeFailure(err, stderr) {
  if (err && (err.killed || err.signal)) {
    return `timed out after ${Math.round(RUN_TIMEOUT_MS / 1000)}s`;
  }
  const line = String(stderr || '')
    .split('\n')
    .map((l) => l.trim())
    .find((l) => l && !/^warning:/i.test(l));
  return line || String((err && err.message) || 'unknown error').split('\n')[0];
}

export function createUsage({ emit }) {
  let latest = loadPersisted();
  let timer = null;
  let running = false;

  function run() {
    // Our own polling is still a Claude Code session. disableAllHooks stops it
    // reporting itself through the hook path, the explicit session id lets the
    // session sources recognise and skip it, and running from a temp dir keeps
    // it out of any project the user actually works in.
    const sessionId = crypto.randomUUID();
    markOwnSession(sessionId);

    return new Promise((resolve) => {
      execFile(
        'claude',
        ['--settings', '{"disableAllHooks":true}', '--session-id', sessionId, '-p', '/usage'],
        {
          timeout: RUN_TIMEOUT_MS,
          cwd: os.tmpdir(),
          env: { ...process.env, CLAUDE_CODE_SKIP_PROMPT_HISTORY: '1' },
          maxBuffer: 1024 * 1024,
          // Closed stdin, not an idle pipe. `claude -p` waits three seconds for
          // input that is never coming, warns about it, and only then starts —
          // three seconds of every run, spent on nothing.
          stdio: ['ignore', 'pipe', 'pipe'],
        },
        (err, stdout, stderr) => {
          if (err) return resolve({ error: describeFailure(err, stderr) });
          resolve({ text: String(stdout) });
        }
      );
    });
  }

  async function refresh() {
    if (running) return latest;
    running = true;
    try {
      const out = await run();
      if (out.error) {
        latest = {
          ...(latest || {}),
          stale: true,
          error: `claude /usage failed: ${out.error}`.slice(0, 120),
        };
      } else {
        const parsed = parseUsage(out.text);
        if (parsed) {
          latest = reconcileUsage(latest, parsed);
          writeState(STATE_NAME, latest);
        } else {
          latest = { ...(latest || {}), stale: true, error: 'could not parse /usage output' };
        }
      }
      emit('claude.usage.update', latest);
    } catch (err) {
      log('US', `usage refresh failed: ${err.message}`);
    } finally {
      running = false;
    }
    return latest;
  }

  return {
    start: () => { refresh(); timer = setInterval(refresh, USAGE_REFRESH_MS); },
    stop: () => clearInterval(timer),
    get: () => latest || { limits: [], error: 'usage not read yet' },
    refresh,
  };
}

File: app/src-tauri/src/daemon/usage.rs
// Real usage, straight from Claude Code. Mirrors daemon/src/usage.js.
//
// `claude -p "/usage"` runs the slash command locally and prints the same
// figures the interactive /usage view shows. Everything here is parsed from
// that text — never invent a denominator.

use crate::daemon::bus::{emit, Emit};
use crate::daemon::config::USAGE_REFRESH_MS;
use crate::daemon::own_sessions::mark_own_session;
use crate::daemon::persist::{read_state, write_state};
use crate::daemon::sessions::store::now_ms;
use chrono::{Local, TimeZone};
use regex::Regex;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::sync::{Arc, Mutex, OnceLock};
use tokio::process::Command;
use tokio::time::Duration;
use uuid::Uuid;

const RUN_TIMEOUT_MS: u64 = 45_000;
const STATE_NAME: &str = "usage";

fn limit_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(r"(?i)^\s*Current\s+(session|week[^:]*):\s*(\d+)%\s*used(?:\s*·\s*resets\s+([^(\n]+?))?\s*(?:\(([^)]+)\))?\s*$")
            .unwrap()
    })
}
fn window_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?i)^\s*Last\s+(\S+)\s*·\s*([\d,]+)\s+requests\s*·\s*([\d,]+)\s+sessions\s*$").unwrap())
}
fn top_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?i)^Top\s+(skills|subagents|MCP servers):\s*(.+)$").unwrap())
}
fn top_item_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"^(.*\S)\s+(\d+%)$").unwrap())
}
fn week_model_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?i)week\s*\(([^)]+)\)").unwrap())
}
fn non_alpha_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"[^a-z]+").unwrap())
}

fn hhmm_local(ms: i64) -> String {
    Local
        .timestamp_millis_opt(ms)
        .single()
        .map(|dt| dt.format("%H:%M").to_string())
        .unwrap_or_default()
}

// Trailing percentage per item; anything that doesn't match that shape is
// kept with an empty value rather than dropped.
fn parse_top_list(rest: &str) -> Vec<Value> {
    rest.split(',')
        .filter_map(|chunk| {
            let item = chunk.trim();
            if item.is_empty() {
                return None;
            }
            Some(match top_item_re().captures(item) {
                Some(caps) => json!({ "name": caps[1].to_string(), "pct": caps[2].to_string() }),
                None => json!({ "name": item, "pct": "" }),
            })
        })
        .collect()
}

fn label_for(kind: &str) -> String {
    let k = kind.to_lowercase();
    if k == "session" {
        return "SESSION".to_string();
    }
    if let Some(caps) = week_model_re().captures(kind) {
        let name = caps[1].trim();
        return if name.to_lowercase().contains("all models") {
            "WEEK · ALL MODELS".to_string()
        } else {
            format!("WEEK · {}", name.to_uppercase())
        };
    }
    kind.to_uppercase()
}

pub fn parse_usage(text: &str, now: i64) -> Option<Value> {
    let mut limits: Vec<Value> = Vec::new();
    let mut windows: Vec<Value> = Vec::new();
    let mut current_idx: Option<usize> = None;
    let mut subscription: Option<String> = None;

    for raw in text.split('\n') {
        let line = raw.trim_end();

        if line.to_lowercase().contains("subscription to power your claude code usage") {
            subscription = Some(line.trim().to_string());
            continue;
        }

        if let Some(caps) = limit_re().captures(line) {
            let kind = caps.get(1).map(|m| m.as_str()).unwrap_or("");
            let used_pct: f64 = caps.get(2).map(|m| m.as_str()).unwrap_or("0").parse().unwrap_or(0.0);
            let detail = caps
                .get(3)
                .map(|m| format!("resets {}", m.as_str().trim()))
                .unwrap_or_default();
            let key = non_alpha_re().replace_all(&kind.to_lowercase(), "-").to_string();
            limits.push(json!({
                "key": key,
                "label": label_for(kind),
                "used": used_pct / 100.0,
                "detail": detail,
            }));
            continue;
        }

        if let Some(caps) = window_re().captures(line) {
            let requests: i64 = caps[2].replace(',', "").parse().unwrap_or(0);
            let sessions: i64 = caps[3].replace(',', "").parse().unwrap_or(0);
            windows.push(json!({
                "window": format!("Last {}", &caps[1]),
                "requests": requests,
                "sessions": sessions,
                "notes": [], "skills": [], "subagents": [], "mcp": [],
            }));
            current_idx = Some(windows.len() - 1);
            continue;
        }

        // indented bullets under a window: behaviours and top skills/etc.
        if let Some(idx) = current_idx {
            let leading_ws = raw.len() - raw.trim_start().len();
            if leading_ws >= 2 && !line.trim().is_empty() {
                let note = line.trim().to_string();
                if let Some(obj) = windows[idx].as_object_mut() {
                    obj.get_mut("notes").unwrap().as_array_mut().unwrap().push(json!(note));
                    if let Some(caps) = top_re().captures(&note) {
                        let key = match caps[1].to_lowercase().as_str() {
                            "skills" => Some("skills"),
                            "subagents" => Some("subagents"),
                            "mcp servers" => Some("mcp"),
                            _ => None,
                        };
                        if let Some(key) = key {
                            obj.insert(key.to_string(), Value::Array(parse_top_list(&caps[2])));
                        }
                    }
                }
            }
        }
    }

    if limits.is_empty() {
        return None;
    }
    Some(json!({
        "updatedTs": now,
        "updatedLabel": format!("updated {} · from claude /usage", hhmm_local(now)),
        "subscription": subscription,
        "limits": limits,
        "windows": windows,
    }))
}

fn same_window(prev: &Value, next: &Value) -> bool {
    let pd = prev.get("detail").and_then(Value::as_str).unwrap_or("");
    let nd = next.get("detail").and_then(Value::as_str).unwrap_or("");
    if pd.is_empty() || nd.is_empty() {
        return true;
    }
    pd == nd
}

// Usage inside a window only ever goes up. A lower number for the same
// window is a stale read (see daemon/src/usage.js), not a refund.
pub fn reconcile_usage(prev: Option<&Value>, next: Value) -> Value {
    let Some(next_limits) = next.get("limits").and_then(Value::as_array).cloned() else {
        return next;
    };
    let before: std::collections::HashMap<String, Value> = prev
        .and_then(|p| p.get("limits"))
        .and_then(Value::as_array)
        .map(|arr| {
            arr.iter()
                .filter_map(|l| l.get("key").and_then(Value::as_str).map(|k| (k.to_string(), l.clone())))
                .collect()
        })
        .unwrap_or_default();

    let limits: Vec<Value> = next_limits
        .into_iter()
        .map(|l| {
            let key = l.get("key").and_then(Value::as_str).unwrap_or("").to_string();
            if let Some(p) = before.get(&key) {
                let p_used = p.get("used").and_then(Value::as_f64).unwrap_or(0.0);
                let l_used = l.get("used").and_then(Value::as_f64).unwrap_or(0.0);
                if p_used > l_used && same_window(p, &l) {
                    let mut merged = l.clone();
                    let detail_l = l.get("detail").and_then(Value::as_str).unwrap_or("");
                    let detail_p = p.get("detail").and_then(Value::as_str).unwrap_or("");
                    merged["used"] = json!(p_used);
                    merged["detail"] = json!(if detail_l.is_empty() { detail_p } else { detail_l });
                    return merged;
                }
            }
            l
        })
        .collect();

    let mut out = next;
    out["limits"] = Value::Array(limits);
    out
}

fn describe_failure(timed_out: bool, stderr: &str) -> String {
    if timed_out {
        return format!("timed out after {}s", RUN_TIMEOUT_MS / 1000);
    }
    stderr
        .lines()
        .map(str::trim)
        .find(|l| !l.is_empty() && !l.to_lowercase().starts_with("warning:"))
        .map(str::to_string)
        .unwrap_or_else(|| "unknown error".to_string())
}

fn load_persisted(state_dir: &Path) -> Option<Value> {
    let saved: Value = read_state(state_dir, STATE_NAME)?;
    let limits = saved.get("limits").and_then(Value::as_array)?;
    if limits.is_empty() {
        return None;
    }
    let at = saved.get("updatedTs").and_then(Value::as_i64).map(hhmm_local).unwrap_or_default();
    let mut out = saved.clone();
    out["stale"] = json!(true);
    if let Some(obj) = out.as_object_mut() {
        obj.remove("error");
    }
    out["updatedLabel"] = json!(format!(
        "last reading{} · from claude /usage",
        if at.is_empty() { String::new() } else { format!(" {at}") }
    ));
    Some(out)
}

struct Inner {
    latest: Option<Value>,
    running: bool,
}

#[derive(Clone)]
pub struct Usage {
    inner: Arc<Mutex<Inner>>,
    emit_tx: Emit,
    state_dir: PathBuf,
}

impl Usage {
    pub fn new(emit_tx: Emit, state_dir: PathBuf) -> Self {
        let latest = load_persisted(&state_dir);
        Usage {
            inner: Arc::new(Mutex::new(Inner { latest, running: false })),
            emit_tx,
            state_dir,
        }
    }

    pub fn get(&self) -> Value {
        self.inner
            .lock()
            .unwrap()
            .latest
            .clone()
            .unwrap_or_else(|| json!({ "limits": [], "error": "usage not read yet" }))
    }

    // No explicit stop(): the refresh loop lives as long as the app process,
    // same as everything else spawned at startup.
    pub fn start(&self) {
        let this = self.clone();
        tokio::spawn(async move {
            let mut ticker = tokio::time::interval(Duration::from_millis(USAGE_REFRESH_MS));
            loop {
                ticker.tick().await;
                this.refresh().await;
            }
        });
    }

    async fn run_claude_usage(&self) -> Result<String, String> {
        let session_id = Uuid::new_v4().to_string();
        mark_own_session(session_id.clone());
        let mut cmd = Command::new("claude");
        cmd.args(["--settings", "{\"disableAllHooks\":true}", "--session-id", &session_id, "-p", "/usage"])
            .current_dir(std::env::temp_dir())
            .env("CLAUDE_CODE_SKIP_PROMPT_HISTORY", "1")
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true);
        let child = cmd.spawn().map_err(|e| e.to_string())?;
        match tokio::time::timeout(Duration::from_millis(RUN_TIMEOUT_MS), child.wait_with_output()).await {
            Ok(Ok(output)) if output.status.success() => Ok(String::from_utf8_lossy(&output.stdout).to_string()),
            Ok(Ok(output)) => Err(describe_failure(false, &String::from_utf8_lossy(&output.stderr))),
            Ok(Err(e)) => Err(e.to_string()),
            Err(_) => Err(describe_failure(true, "")),
        }
    }

    pub async fn refresh(&self) -> Value {
        {
            let mut inner = self.inner.lock().unwrap();
            if inner.running {
                return inner.latest.clone().unwrap_or_else(|| json!({ "limits": [] }));
            }
            inner.running = true;
        }
        let result = self.run_claude_usage().await;
        {
            let mut inner = self.inner.lock().unwrap();
            match result {
                Err(err) => {
                    let mut latest = inner.latest.clone().unwrap_or_else(|| json!({}));
                    latest["stale"] = json!(true);
                    let msg: String = format!("claude /usage failed: {err}").chars().take(120).collect();
                    latest["error"] = json!(msg);
                    inner.latest = Some(latest);
                }
                Ok(text) => match parse_usage(&text, now_ms()) {
                    Some(parsed) => {
                        let reconciled = reconcile_usage(inner.latest.as_ref(), parsed);
                        inner.latest = Some(reconciled.clone());
                        write_state(&self.state_dir, STATE_NAME, &reconciled);
                    }
                    None => {
                        let mut latest = inner.latest.clone().unwrap_or_else(|| json!({}));
                        latest["stale"] = json!(true);
                        latest["error"] = json!("could not parse /usage output");
                        inner.latest = Some(latest);
                    }
                },
            }
            inner.running = false;
        }
        let value = self.inner.lock().unwrap().latest.clone().unwrap_or_else(|| json!({ "limits": [] }));
        emit(&self.emit_tx, "claude.usage.update", value.clone());
        value
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_session_and_week_limits() {
        let text = concat!(
            "Current session: 11% used · resets Jul 30 at 5:19am (America/New_York)\n",
            "Current week (Sonnet): 42% used\n",
            "Last 24h · 579 requests · 8 sessions\n",
            "  Top skills: /webapp-testing 4%, /frontend-design 1%\n",
        );
        let parsed = parse_usage(text, 0).expect("should parse");
        let limits = parsed["limits"].as_array().unwrap();
        assert_eq!(limits.len(), 2);
        assert_eq!(limits[0]["key"], "session");
        assert!((limits[0]["used"].as_f64().unwrap() - 0.11).abs() < 1e-9);
        assert_eq!(limits[1]["label"], "WEEK · SONNET");
        let windows = parsed["windows"].as_array().unwrap();
        assert_eq!(windows[0]["requests"], 579);
        assert_eq!(windows[0]["skills"][0]["pct"], "4%");
    }

    #[test]
    fn reconcile_keeps_higher_reading_same_window() {
        let prev = json!({ "limits": [{ "key": "session", "used": 0.5, "detail": "resets 5pm" }] });
        let next = json!({ "limits": [{ "key": "session", "used": 0.1, "detail": "resets 5pm" }] });
        let out = reconcile_usage(Some(&prev), next);
        assert!((out["limits"][0]["used"].as_f64().unwrap() - 0.5).abs() < 1e-9);
    }

    #[test]
    fn reconcile_takes_real_rollover() {
        let prev = json!({ "limits": [{ "key": "session", "used": 0.5, "detail": "resets 5pm" }] });
        let next = json!({ "limits": [{ "key": "session", "used": 0.05, "detail": "resets 9pm" }] });
        let out = reconcile_usage(Some(&prev), next);
        assert!((out["limits"][0]["used"].as_f64().unwrap() - 0.05).abs() < 1e-9);
    }
}
